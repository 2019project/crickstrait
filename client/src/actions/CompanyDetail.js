import {
  COMPANY_DETAIL,
  COMPANY_DETAIL_BY_ID,
  COMPANY_DATES_BY_ID,
  OHLC_CHART,
  GET_SIMILAR_TABLE,
  GET_DROP_DOWN,
} from "./Types";
import { startLoading, stopLoading } from "./LoadingAction";
import axios from "axios";

// getting company details[NIKHIL]
export const getCompanyDetail = () => dispatch => {
  try {
    return axios
      .get("http://localhost:2001/api/companydetail/all")
      .then(res => {
        dispatch({
          type: COMPANY_DETAIL,
          payload: res.data.data,
        });
      });
  } catch (err) {
    console.log(err);
  }
};

// getting company details by id [NIKHIL]
export const getCompanyDetailById = id => dispatch => {
  console.log("get companydetails by id from actions", id);
  try {
    return axios
      .get("http://localhost:2001/api/companydetail/" + id)
      .then(res => {
        dispatch({
          type: COMPANY_DETAIL_BY_ID,
          payload: res.data.data,
        });
        console.log(res.data.data.result.sector);
        let sector = {
          sector: res.data.data.result.sector,
        };
        console.log("obj sector from action", sector);

        //To call another action within the action we call the function in the dipatch.
        // calling another action
        dispatch(getSimilarTable(sector));

        console.log("from then");
      });
  } catch (err) {
    console.log(err);
  }
};

//GETTING DATA FOR DROPDOWN
export const getDropDownData = sector => dispatch => {
  console.log("actions of dropdown", sector);
  try {
    return axios
      .post("http://localhost:2001/api/dropdown", sector)
      .then(res => {
        dispatch({
          type: GET_DROP_DOWN,
          payload: res.data.data,
        });
        console.log("from then of drop down action");
      });
  } catch (err) {
    console.log(err);
  }
};

//getting analysis
export const getSimilarTable = sector => dispatch => {
  // const sector = requset.body.sector;
  console.log("get analysis by id from actions", sector);
  try {
    return axios
      .post("http://localhost:2001/api/analysis/analysis", sector)
      .then(res => {
        dispatch({
          type: GET_SIMILAR_TABLE,

          payload: res.data.data,
        });
        console.log("from then of similar table");
        dispatch(getDropDownData(sector));
      });
  } catch (err) {
    console.log(err);
  }
};

// Fetching OHLC Chart for company [Bhavana]
export const getOhlcChart = id => dispatch => {
  dispatch(startLoading());
  try {
    return axios.get("http://localhost:5000/companyindices/" + id).then(res => {
      dispatch(stopLoading());
      dispatch({
        type: OHLC_CHART,

        payload: res.data,
      });
    });
  } catch (err) {
    dispatch(startLoading());
    console.log(err);
  }
};

// downloading ohlc data for company[piyush]

export const downloadOhlcDataCompany = ohlc => {
  try {
    return axios.get(
      "http://localhost:2001/api/companydetail/downloadohlc/" + ohlc
    );
  } catch (err) {
    console.log(err);
  }
};

// getting company dates by id
export const getCompanyDatesById = id => dispatch => {
  console.log("get companydetails by id from actions", id);
  try {
    return axios
      .get("http://localhost:2001/api/companydetail/financial/" + id)
      .then(res => {
        dispatch({
          type: COMPANY_DATES_BY_ID,
          payload: res.data.data,
        });
        //  let sector = {
        //   sector: res.data.data.result.sector
        // };
        //  dispatch(getSimilarTable(sector));
      });
  } catch (err) {
    console.log(err);
  }
};
