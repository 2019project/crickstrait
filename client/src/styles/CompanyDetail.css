/* FOR MOBILE  */

#company-detail-grid-container {
  padding: 10px 10px 10px 10px;
  grid-template-rows: 1fr;
  display: grid;
  grid-template-areas:
    "profile"
    "recommendation";
  grid-template-columns: 1fr;
  font-size: 10px;

  text-align: justify;
}

#company-detail-recommendation {
  grid-area: recommendation;
}
#company-detail-profile {
  grid-area: profile;
}
#company-detail-profile > h3 > span {
  /* color: red; */
  color: #707070;
  font-size: 20px;
}
#h3 {
  color: red;
}

/* FOR DESKTOP  */
@media screen and (min-width: 480px) {
  #company-detail-grid-container {
    padding: 20px 50px 20px 50px;
    grid-template-rows: 1fr;
    display: grid;
    grid-template-areas: "profile recommendation";
    /* "chart chart"; */
    grid-template-columns: 1fr 1fr;
    /* overflow-x: scroll; */
    /* overflow-y: scroll; */
  }

  #company-detail-profile > h3 {
    /* color: #707070; */
    font-size: 20px;
    text-align: justify;
  }

  /* .profile {
} */

  #company-detail-profile {
    grid-area: profile;
    /* border: 1px solid; */
    top: 50%;
    bottom: 0;

    font-size: 100px;
  }

  #company-detail-recommendation {
    grid-area: recommendation;
    text-align: center;
    /* height: 102px; */
    /* border: 2px solid black; */
  }

  #img_gauge {
    width: 100px;
  }

  #test {
    border-left: 1px solid #cacaca;
    height: 500px;
    margin-top: 100px;
  }
}
