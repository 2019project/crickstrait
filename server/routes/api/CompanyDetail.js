const express = require("express");
const router = express.Router();
var result;

const MongoClient = require("mongodb").MongoClient;
const assert = require("assert");

// Connection URL
const url = "mongodb://localhost:27017";

// Database Name
const dbName = "stocks";
const stocksData = require("../../model/stocksModel");

//Use connect method to connect to the server
MongoClient.connect(url, function(err, client) {
  assert.equal(null, err);
  console.log("Connected successfully to server");

  const db = client.db(dbName);
  router.get("/:id", async (req, res, next) => {
    try {
      let id = req.params.id;
      // console.log("printing id from api all", id);
      var collection = db.collection("stocks_data_2");
      // hardcoding dummy data
      var dummy_date = [
        "2009-09-30",
        "2009-12-31",
        "2010-03-31",
        "2010-06-30",
        "2010-09-30",
        "2010-12-31",
        "2011-03-31",
        "2011-06-30",
        "2011-09-30",
        "2011-12-31"
      ]; //variable
      collection.findOne({ ticker_id: +id }, { sector: 1 }, function(
        err,
        result
      ) {
        // console.log(result);
        // var second_last_date = result.slice(-2)[0];
        // console.log(second_last_date);

        // console.log(result);
        var balancesheet = [];
        var cashflow = [];
        var profitandloss = [];
        var ratios = [];
        var company_details = [];
        // var company_id = [result.ticker_id];
        //will contain company profile
        // var company_profile = [result.profile];
        // var company_sector = [result.sector];
        // console.log("result", company_profile);

        // console.log("printing only id from result", result);

        // Fetching the object using dates for balance sheet
        // for (i of dummy_date) {
        //   console.log("new i", i);
        //   var balance = {};
        //   var flow = {};
        //   var ratio = {};
        //   var pnl = {};
        //   var details = {};
        //   balance["ticker_dates"] = i;
        //   flow["ticker_dates"] = i;
        //   pnl["ticker_dates"] = i;
        //   ratio["ticker_dates"] = i;
        //   // balance["ticker_dates"] = i;
        //   console.log("new bal", balance);
        //   balance["Cash and Cash Equivalents"] = result.ticker_dates[i][
        //     "Cash and Cash Equivalents"
        //   ]
        //     ? result.ticker_dates[i]["Cash and Cash Equivalents"]
        //     : "-";
        //   console.log("inside", balance);
        //   balance["Current Assets"] = result.ticker_dates[i]["Current Assets"];
        //   balance["Total Assets"] = result.ticker_dates[i]["Total Assets"];

        //   balance["Accounts Payable"] =
        //     result.ticker_dates[i]["Accounts Payable"];

        //   balance["Receivables"] = result.ticker_dates[i]["Receivables"];

        //   balance["Total Liabilities"] =
        //     result.ticker_dates[i]["Total Liabilities"];

        //   balance["Current Liabilities"] =
        //     result.ticker_dates[i]["Current Liabilities"];

        //   balance["Preferred Equity"] =
        //     result.ticker_dates[i]["Preferred Equity"];

        //   balance["Equity Before Minorities"] =
        //     result.ticker_dates[i]["Equity Before Minorities"];

        //   balance["Minorities Interest"] = result.ticker_dates[i]["Minorities"];

        //   balance["Noncurrent Liabilities"] =
        //     result.ticker_dates[i]["Total Noncurrent Liabilities"];
        //   balancesheet.push(balance);

        //   //cashflow
        //   flow["Cash From Operating Activities"] =
        //     result.ticker_dates[i]["Cash From Operating Activities"];

        //   flow["Cash From Investing Activities"] =
        //     result.ticker_dates[i]["Cash From Investing Activities"];

        //   flow["Cash From Financing Activities"] =
        //     result.ticker_dates[i]["Cash From Financing Activities"];

        //   flow["EBITDA"] = result.ticker_dates[i]["EBITDA"];

        //   flow["Net Change in Cash"] =
        //     result.ticker_dates[i]["Net Change in Cash"];

        //   flow["Net PP&E"] = result.ticker_dates[i]["Net PP&E"];

        //   flow["Dividends"] = result.ticker_dates[i]["Dividends"];
        //   cashflow.push(flow);

        //   //PROFIT AND LOSS
        //   pnl["Revenues"] = result.ticker_dates[i]["Revenues"];
        //   pnl["EBIT"] = result.ticker_dates[i]["EBIT"];

        //   pnl["Net Profit"] = result.ticker_dates[i]["Net Profit"];

        //   pnl["Revenues"] = result.ticker_dates[i]["Revenues"];
        //   profitandloss.push(pnl);

        //   //calculating ratios
        //   ratio["Current Ratio"] =
        //     result.ticker_dates[i]["Current Assets"] /
        //     result.ticker_dates[i]["Current Liabilities"];

        //   ratio["Liabilities To Equity"] =
        //     result.ticker_dates[i]["Total Liabilities"] /
        //     result.ticker_dates[i]["Total Equity"];

        //   ratio["Debt To Asset"] =
        //     result.ticker_dates[i]["Total Assets"] /
        //     result.ticker_dates[i]["Total Liabilities"];
        //   ratios.push(ratio);
        // }
        // var company_details = [];
        // details["ticker_id"] = result.ticker_id;
        // details["sector"] = result.sector;
        // details["industry"] = result.industry;

        // details["ticker_name"] = result.ticker_name;
        // details["employess"] = result.employess;

        // details["profile"] = result.profile;

        // details["company_name"] = result.company_name;
        // company_details.push(details);

        // console.log("Balance sheet", balancesheet);
        // console.log("cashflow sheet", cashflow);
        // console.log("pnl", profitandloss);

        // console.log("ratios", ratios);
        // console.log(company_details);

        if (!result) {
          return res.status(400).send({ message: "No data found" });
        } else {
          if (err) throw err;
          res.status(200).json({
            status: 200,
            data: {
              // balancesheet: balancesheet,
              // cashflow: cashflow,
              // profitandloss: profitandloss,
              // ratios: ratios,
              // ticker_id: result.ticker_id,
              // profile: result.profile,
              // industry: result.industry,
              // company_name: result.company_name,
              // employess: result.employess,
              // ticker_name: result.ticker_name,
              // sector: result.sector
              result
            },

            message: "Retrieved data from company detail successfully"
          });
        }
      });
    } catch (err) {
      next(err);
    }
  });
});

// getting cash flow

// router.get("/:id", async (req, res, next) => {
//   try {
//     let id = req.params.id;
//     console.log("printing id from api all", id);
//     var collection = db.collection("stocks");
//     collection.findOne({ _id: +id }, function(err, result) {
//       print(result);
//       if (!result) {
//         return res.status(400).send({ message: "No data found" });
//       } else {
//         if (err) throw err;
//         res.status(200).json({
//           status: 200,
//           data: result,
//           message: "Retrieved news Successfully"
//         });
//       }
//     });
//   } catch (err) {
//     next(err);
//   }
// });

//for analysis

//new financials
router.get("/financial/:id", async (req, res, next) => {
  try {
    let id = req.params.id;
    //variable
    // stocksData.findOne({ticker_name: 'AAPL',"ticker_dates":{'$elemMatch': {date: new Date("2019-03-31")}}}, function(err, result) {
    stocksData.aggregate(
      [
        { $unwind: "$ticker_dates" },
        {
          $match: {
            // ticker_name : "AAPL"
            ticker_id: +id
            // ,
            // 'ticker_dates.date' : {
            //   $lte : new Date("2019-06-31"),
            //   $gte :  new Date("2018-03-25")
            // }
          }
        },
        {
          $project: {
            ticker_dates: 1,
            // ,"ticker_dates.Market Capitalisation": 1,

            quarter: {
              $cond: [
                {
                  $and: [
                    { $eq: [{ $month: "$ticker_dates.date" }, 3] },
                    { $lte: [{ $dayOfMonth: "$ticker_dates.date" }, 31] },
                    { $gt: [{ $dayOfMonth: "$ticker_dates.date" }, 25] }
                  ]
                },
                "first",
                {
                  $cond: [
                    {
                      $and: [
                        { $eq: [{ $month: "$ticker_dates.date" }, 6] },
                        { $lte: [{ $dayOfMonth: "$ticker_dates.date" }, 30] },
                        { $gt: [{ $dayOfMonth: "$ticker_dates.date" }, 25] }
                      ]
                    },
                    "second",
                    {
                      $cond: [
                        {
                          $and: [
                            { $eq: [{ $month: "$ticker_dates.date" }, 9] },
                            {
                              $lte: [{ $dayOfMonth: "$ticker_dates.date" }, 30]
                            },
                            { $gt: [{ $dayOfMonth: "$ticker_dates.date" }, 25] }
                          ]
                        },
                        "third",
                        {
                          $cond: [
                            {
                              $and: [
                                { $eq: [{ $month: "$ticker_dates.date" }, 12] },
                                {
                                  $lte: [
                                    { $dayOfMonth: "$ticker_dates.date" },
                                    31
                                  ]
                                },
                                {
                                  $gt: [
                                    { $dayOfMonth: "$ticker_dates.date" },
                                    25
                                  ]
                                }
                              ]
                            },
                            "fourth",
                            "fifth"
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          }
        },
        { $match: { quarter: { $ne: "fifth" } } },
        { $sort: { "ticker_dates.date": -1 } },
        {
          $group: {
            _id: { quarter: "$quarter" },
            results: { $push: "$ticker_dates" }
          }
        }
      ],
      function(err, result) {
        //   console.log("start");
        // console.log("end");

        if (!result) {
          return res.status(400).json({
            status: 400,
            data: result,
            message: "Retrieved dates Successfully"
          });
        } else {
          if (err) throw err;
          res.status(200).json({
            status: 200,
            data: result,
            message: "Retrieved dates Successfully"
          });
        }
      }
    );
  } catch (err) {
    console.log(err);
    next(err);
  }
});

//closing the connect method

router.post("/analysis", async (req, res, next) => {
  console.log("analysis called");
  try {
    console.log("re body", req.body.sector);
    const sector = req.body.sector;

    let result = await stocksData.find({ sector: { $in: sector } }).limit(5);
    // console.log(result);

    if (result < 0) {
      res.status(400).json({
        status: 400,
        data: result,
        message: "No news Found"
      });
    } else {
      res.status(200).json({
        status: 200,
        data: result,
        message: "Retrieved all news Successfully"
      });
    }
  } catch (err) {
    next(err);
  }
});
module.exports = router;
