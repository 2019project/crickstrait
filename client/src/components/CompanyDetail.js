import React, { Component } from "react";
import { getCompanyDetail } from "../actions/CompanyDetail";
import { connect } from "react-redux";
//importing css file
import "../styles/CompanyDetail.css";

export class CompanyDetail extends Component {
  componentDidMount() {
    console.log("component mounted");
    this.props.getCompanyDetail();
  }
  render() {
    return (
      <div>
        <div id="company-detail-grid-container">
          {this.props.company.map(company => (
            <div id="company-detail-profile">
              <h3>
                Sector :<span>{company.sector}</span>
              </h3>

              <h3>
                Market Cap: <span> {company.marketcap}</span>
              </h3>
              <h3>
                Industry: <span> {company.industry}</span>
              </h3>
              <h3>
                Employees: <span> {company.employees}</span>
              </h3>
              <h3>
                <span>{company.bio}</span>{" "}
              </h3>
            </div>
          ))}
          {/* recommendation */}

          <div id="company-detail-recommendation">
            {this.props.company.map(company => (
              <h3> {company.sector}</h3>
            ))}
            <div id="test"></div>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  company: state.CompanyDetailReducer.company,
});
export default connect(
  mapStateToProps,
  { getCompanyDetail }
)(CompanyDetail);
