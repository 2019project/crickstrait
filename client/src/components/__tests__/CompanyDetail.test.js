import React from 'react'
// import { Home } from '../Home'
import { shallow, mount } from 'enzyme'
import { CompanyDetail } from '../CompanyDetail'

const getCompanyDetailById = jest.fn()
const getOhlcChart = jest.fn()

const company = [
  {
    sector: 'technology',
    industry: 'Computer',
    employees: '123456',
    profile: 'profile'
  }
]
const balancesheet = [{}]

const wrapper = shallow(
  <CompanyDetail
    getCompanyDetailById={getCompanyDetailById}
    company={company}
    match={{ params: { id: 1 } }}
    getOhlcChart={getOhlcChart}
  />
)

describe('Testing Company Detail Component', () => {
  //   afterEach(() => {
  //     jest.clearAllMocks()

  it('should mount the component', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('should display the sector', () => {
    expect(wrapper.find('#company-detail-sector').text()).toBe('technology')
  })
  it('should display the industry', () => {
    expect(wrapper.find('#company-detail-industry').text()).toBe('computer')
  })
  it('should display the employees', () => {
    expect(wrapper.find('#company-detail-employees').text()).toBe('123456')
  })
  it('should display the profile', () => {
    expect(wrapper.find('#company-detail-profile').text()).toBe('profile')
  })
})
