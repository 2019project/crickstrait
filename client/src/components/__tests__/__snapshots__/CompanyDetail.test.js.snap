// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Company Detail Component should mount the component 1`] = `
<div>
  <Connect(CompanyDetailSecondaryNavbar) />
  <p>
    Loading..
  </p>
  <div
    style={
      Object {
        "border": "1px solid #cacaca",
        "margin": "auto",
        "marginBottom": "20px",
        "width": "97%",
      }
    }
  >
    <iframe
      style={
        Object {
          "border": "none",
          "height": "550px",
          "outline": "none",
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;
